File: ios/Packages/Presentation/Sources/Presentation/Screen/Calculator/CalculatorKey.swift
/// 계산기 키 정의 (Android `CalculatorKey` 대응)
enum CalculatorKey: Hashable {
    case history
    case clear
    case parenthesis
    case dot
    case delete
    case calculate
    case number(String)
    case operatorKey(displayText: String, inputValue: String)
}

let calculatorKeyRows: [[CalculatorKey]] = [
    [.history, .clear, .parenthesis, .operatorKey(displayText: "÷", inputValue: "÷")],
    [.number("7"), .number("8"), .number("9"), .operatorKey(displayText: "×", inputValue: "×")],
    [.number("4"), .number("5"), .number("6"), .operatorKey(displayText: "−", inputValue: "-")],
    [.number("1"), .number("2"), .number("3"), .operatorKey(displayText: "+", inputValue: "+")],
    [.dot, .number("0"), .delete, .calculate],
]

extension CalculatorKey {
    func toIntent() -> CalculatorIntent? {
        switch self {
        case .history:
            nil

        case .clear:
            .clear

        case .parenthesis:
            .input(.parenthesis)

        case .dot:
            .input(.dot)

        case .delete:
            .delete

        case .calculate:
            .calculate

        case let .number(value):
            .input(.number(value))

        case let .operatorKey(_, inputValue):
            .input(.operator(inputValue))
        }
    }
}

File: presentation/src/main/java/com/ahn/presentation/ui/screen/calculator/CalculatorKey.kt
package com.ahn.presentation.ui.screen.calculator

internal sealed interface CalculatorKey {
    data object History : CalculatorKey

    data object Clear : CalculatorKey

    data object Parenthesis : CalculatorKey

    data object Dot : CalculatorKey

    data object Delete : CalculatorKey

    data object Calculate : CalculatorKey

    data class Number(val value: String) : CalculatorKey

    data class Operator(
        val displayText: String,
        val inputValue: String,
    ) : CalculatorKey
}

internal val calculatorKeyRows = listOf(
    listOf(
        CalculatorKey.History,
        CalculatorKey.Clear,
        CalculatorKey.Parenthesis,
        CalculatorKey.Operator(displayText = "÷", inputValue = "÷"),
    ),
    listOf(
        CalculatorKey.Number("7"),
        CalculatorKey.Number("8"),
        CalculatorKey.Number("9"),
        CalculatorKey.Operator(displayText = "×", inputValue = "×"),
    ),
    listOf(
        CalculatorKey.Number("4"),
        CalculatorKey.Number("5"),
        CalculatorKey.Number("6"),
        CalculatorKey.Operator(displayText = "−", inputValue = "-"),
    ),
    listOf(
        CalculatorKey.Number("1"),
        CalculatorKey.Number("2"),
        CalculatorKey.Number("3"),
        CalculatorKey.Operator(displayText = "+", inputValue = "+"),
    ),
    listOf(
        CalculatorKey.Dot,
        CalculatorKey.Number("0"),
        CalculatorKey.Delete,
        CalculatorKey.Calculate,
    ),
)

internal fun CalculatorKey.toIntent(): CalculatorContract.Intent? =
    when (this) {
        CalculatorKey.History -> null

        CalculatorKey.Clear -> CalculatorContract.Intent.Clear

        CalculatorKey.Parenthesis ->
            CalculatorContract.Intent.Input(CalculatorToken.Parenthesis)

        CalculatorKey.Dot -> CalculatorContract.Intent.Input(CalculatorToken.Dot)

        CalculatorKey.Delete -> CalculatorContract.Intent.Delete

        CalculatorKey.Calculate -> CalculatorContract.Intent.Calculate

        is CalculatorKey.Number ->
            CalculatorContract.Intent.Input(CalculatorToken.Number(value))

        is CalculatorKey.Operator ->
            CalculatorContract.Intent.Input(CalculatorToken.Operator(inputValue))
    }
